File: Java/ECommerceSystem.java
import java.util.ArrayList;
import java.util.List;

public class ECommerceSystem {
    
    public static void checkout(Customer customer, Cart cart) {
        // Check if cart is empty
        if (cart.isEmpty()) {
            System.out.println("Error: Cart is empty");
            return;
        }
        
        // Check for expired products and stock availability
        List<Shippable> shippableItems = new ArrayList<>();
        
        for (CartItem item : cart.getItems()) {
            Product product = item.getProduct();
            
            // Check if product is expired
            if (product.isExpired()) {
                System.out.println("Error: " + product.getName() + " is expired");
                return;
            }
            
            // Check if product is out of stock
            if (!product.isAvailable(item.getQuantity())) {
                System.out.println("Error: " + product.getName() + " is out of stock");
                return;
            }
            
            // Collect shippable items
            if (product.requiresShipping() && product instanceof Shippable) {
                // Add multiple instances for quantity
                for (int i = 0; i < item.getQuantity(); i++) {
                    shippableItems.add((Shippable) product);
                }
            }
        }
        
        // Calculate costs
        double subtotal = cart.getSubtotal();
        double shippingFee = ShippingService.calculateShippingFee(shippableItems);
        double totalAmount = subtotal + shippingFee;
        
        // Check if customer has sufficient balance
        if (!customer.hasBalance(totalAmount)) {
            System.out.println("Error: Customer's balance is insufficient");
            return;
        }
        
        // Process payment
        customer.deductBalance(totalAmount);
        
        // Update product quantities
        for (CartItem item : cart.getItems()) {
            Product product = item.getProduct();
            product.setQuantity(product.getQuantity() - item.getQuantity());
        }
        
        // Send shipment notice
        if (!shippableItems.isEmpty()) {
            ShippingService.ship(shippableItems);
        }
        
        // Print checkout receipt
        System.out.println("** Checkout receipt **");
        for (CartItem item : cart.getItems()) {
            System.out.println(item.getQuantity() + "x " + item.getProduct().getName() + " " + (int)item.getTotalPrice());
        }
        System.out.println("----------------------");
        System.out.println("Subtotal " + (int)subtotal);
        System.out.println("Shipping " + (int)shippingFee);
        System.out.println("Amount " + (int)totalAmount);
        System.out.println("Customer balance after payment: " + customer.getBalance());
        
        // Clear cart
        cart.clear();
    }
}

File: python/ecommerce_system.py
from customer import Customer
from cart import Cart
from shipping_service import ShippingService
from shippable import Shippable

class ECommerceSystem:
    @staticmethod
    def checkout(customer, cart):
        if cart.is_empty():
            print("Error: Cart is empty")
            return

        shippable_items = []
        for item in cart.get_items():
            product = item.get_product()

            if product.is_expired():
                print(f"Error: {product.get_name()} is expired")
                return

            if not product.is_available(item.get_quantity()):
                print(f"Error: {product.get_name()} is out of stock")
                return

            if product.requires_shipping() and isinstance(product, Shippable):
                for _ in range(item.get_quantity()):
                    shippable_items.append(product)

        subtotal = cart.get_subtotal()
        shipping_fee = ShippingService.calculate_shipping_fee(shippable_items)
        total_amount = subtotal + shipping_fee

        if not customer.has_balance(total_amount):
            print("Error: Customer's balance is insufficient")
            return

        customer.deduct_balance(total_amount)

        for item in cart.get_items():
            product = item.get_product()
            product.set_quantity(product.get_quantity() - item.get_quantity())

        if shippable_items:
            ShippingService.ship(shippable_items)

        print("** Checkout receipt **")
        for item in cart.get_items():
            print(f"{item.get_quantity()}x {item.get_product().get_name()} {int(item.get_total_price())}")
        print("----------------------")
        print(f"Subtotal {int(subtotal)}")
        print(f"Shipping {int(shipping_fee)}")
        print(f"Amount {int(total_amount)}")
        print(f"Customer balance after payment: {customer.get_balance()}")

        cart.clear()
